Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm

```diff
@@ -2154,35 +2156,66 @@ static void fulfillRequestWithKeyData(AVAssetResourceLoadingRequest *request, Ar
         return true;
     }
 
-#if ENABLE(LEGACY_ENCRYPTED_MEDIA)
     if (scheme == "clearkey"_s) {
         String keyID = [[[avRequest request] URL] resourceSpecifier];
         auto encodedKeyId = PAL::TextCodecUTF8::encodeUTF8(keyID);
         auto initData = SharedBuffer::create(WTFMove(encodedKeyId));
 
+#if ENABLE(ENCRYPTED_MEDIA)
+        if (requestKeyHandleForKeyID(keyID, avRequest))
+            return true;
+#endif
+
+#if ENABLE(LEGACY_ENCRYPTED_MEDIA)
         auto keyData = player->cachedKeyForKeyId(keyID);
         if (keyData) {
             fulfillRequestWithKeyData(avRequest, keyData.get());
             return false;
         }
+#endif
+
+#if ENABLE(ENCRYPTED_MEDIA)
+        player->initializationDataEncountered(scheme, initData->tryCreateArrayBuffer());
+        setWaitingForKey(true);
+#endif
 
+#if ENABLE(LEGACY_ENCRYPTED_MEDIA)
         player->keyNeeded(initData);
+#endif
 
         if (!player->shouldContinueAfterKeyNeeded())
             return false;
 
         m_keyURIToRequestMap.set(keyID, avRequest);
         return true;
     }
-#endif
-#endif
+#endif // ENABLE(LEGACY_ENCRYPTED_MEDIA) || ENABLE(ENCRYPTED_MEDIA)
 
     auto resourceLoader = WebCoreAVFResourceLoader::create(this, avRequest);
     m_resourceLoaderMap.add((__bridge CFTypeRef)avRequest, resourceLoader.copyRef());
     resourceLoader->startLoading();
     return true;
 }
 
+bool MediaPlayerPrivateAVFoundationObjC::requestKeyHandleForKeyID(const String& keyID, RetainPtr<AVAssetResourceLoadingRequest> avRequest)
+{
+    auto* clearKeyInstance = dynamicDowncast<CDMInstanceClearKey>(m_cdmInstance.get());
+    if (!clearKeyInstance)
+        return false;
+
+    auto encodedKeyId = PAL::TextCodecUTF8::encodeUTF8(keyID);
+
+    auto promise = clearKeyInstance->getKeyHandleValue(encodedKeyId)->whenSettled(RunLoop::main(), [avRequest = WTFMove(avRequest)] (const auto& result) mutable {
```

Comment:
unused return value, there's no need to store it. `auto promise = ` is unnecessary